### PR TITLE
s3_redundant selfhealing

### PR DESCRIPTION
**Pull request by muhamadazmy**
Implement self-healing for s3_redundant. It should handle minio vm failure and namespaces failure for both tlog and data.

Fixes:
https://github.com/threefoldtech/0-templates/issues/155
https://github.com/threefoldtech/0-templates/issues/154
https://github.com/threefoldtech/0-templates/issues/153
https://github.com/threefoldtech/0-templates/issues/152

### CODE REVIEW

File: templates/s3/s3.py [hunk 1]
```diff
@@ -182,7 +212,33 @@ def update_state():
...
+        master_connection = namespace_connection_info(namespace)
+        self.data['master']['address'] = master_connection
+
+        return {
+            'address': master_connection,
+            'namespace': self._tlog_namespace,
+        }
+
     def install(self):
+        def _get_master_info():
```
Comment (zaibon) [inline templates/s3/s3.py]: This is duplicate code of the method line 215

File: templates/s3/s3.py [hunk 2]
```diff
@@ -442,7 +506,7 @@ def _deploy_minio_vm(self):
         }
         vm_data = {
             'cpu': 2,
-            'memory': 4000,
+            'memory': 1000,
```
Comment (zaibon) [inline templates/s3/s3.py]: don't change this value

File: templates/s3/s3.py [hunk 3]
```diff
@@ -473,7 +536,7 @@ def _deploy_minio_vm(self):
             t = vm.schedule_action('install')
             t.wait()
             if t.state != 'ok':
-                vm.delete(wait=True, timeout=60, die=False)
+                vm.schedule_action('uninstall')(wait=True, timeout=60, die=False)
```
Comment (zaibon) [inline templates/s3/s3.py]: we still need to delete the service after uninstalling it

File: templates/s3/s3.py [hunk 4]
```diff
@@ -522,7 +585,7 @@ def _deploy_namespaces(self, nr_namepaces, name,  size, storage_type, password,
                     deployed_nr_namespaces += 1
 
                     # update amount of ressource so the next iteration of the loop will sort the list of nodes properly
-                    nodes[nodes.index(node)]['used_resources'][storage_key] += size
+                    # nodes[nodes.index(node)]['used_resources'][storage_key] += size
```
Comment (zaibon) [inline templates/s3/s3.py]: why is this commented ?

File: templates/s3/s3.py [hunk 5]
```diff
@@ -616,6 +731,7 @@ def namespace_connection_info(namespace):
 
 
 def sort_by_less_used(nodes, storage_key):
+    return nodes
```
Comment (zaibon) [inline templates/s3/s3.py]: development hack ?

File: templates/s3/s3.py [hunk 6]
```diff
@@ -85,6 +90,17 @@ def _update_url(self):
 
         return self.data['minioUrls']
 
+    def _get_namespace_by_address(self, address):
```
[the remark this answers is not in the document]
Comment (serboctor) [reply templates/s3/s3.py]: @andrewayoub

**Pull request merged**